### DISCUSSION TITLE
2.4.0 : Dynamic Task Mapping - Assigning multiple parameters to a PythonOperator

### DISCUSSION THREAD

**Discussion by EssKayz** [apache/airflow#26833] [Q&A]
### Description

After the new implementation of "Assigning multiple parameters to a non-TaskFlow operator" we tried using this for our PythonOperators, but couldn't get it to work. This is seemingly because a PythonOperator calls the callable function using `op_kwargs`, and as such anything defined in `expand_kwargs` doesn't get routed to the python_callable.

### Use case/motivation

We use Airflow quite heavily using PythonOperators, where data flows through XCom -> it would be really nice to be able to pass over dictionaries from XCom through op_kwargs using the dynamic task mapping. Currently we are abusing the PythonOperator's "template_dict" for this purpose, by doing:
```
PythonOperator.partial(...).expand(
            templates_dict=XComArg(task)
)
```

This way we can currently get a dictionary from XCom into the python_callable, and then parsing that into keywords after the fact.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by uranusjr** [comment]
Can you provide a concrete example? It sounds to me that taskflow is perfect for your use case instead.

**Comment by EssKayz** [comment]
We tried taskflow aswell, but with taskflow we ran into the problem that we cannot dynamically give task_id to the generated tasks. We are using Airflow for an ETL migration process, so we have automagic task generation for data exports, transformations, filtering and imports.

Concrete example pseudocode:
```
for each table in database:
    generate following tasks, with table.table_name included in the task_ids (50+ tasks with the same name would get extremely confusing)
        task that exports data as list of json from database, and splits it into 1...n S3 bucket files,
        and then another task that dynamically generated task, that takes one of the split files, and performs a python_function on it. After this - collect all the converted objects from the dynamically generated tasks, and combine them into one S3 bucket file.
        
```

So essentially what we do currently, is have the splitting task return {'bucket_key': '....'} as the XCOM value, and then that is in turn passed over to PythonOperator as templates_dict 
-> that in turn then gets passed to another function using function(**templates_dict)
Which finally means that our processing function gets the keyword argument bucket_key from the expanded XCOM property

**Comment by uranusjr** [comment]
But you can’t dynamically provide a task_id to PythonOperator either…? I think I am mising something very important here. Again, a concrete example would be best.

**Comment by EssKayz** [comment]
I sadly am not allowed to post direct code due to company policies, but how we deal with that is, we for example have a function  that does roughly the following for all tables in customer's database: 
```
def gen_multiprocessed_task(gen_task_id, processing_func, ...):
task = PythonOperator(
            task_id=f'split_{gen_task_id}_data',
            python_callable=split_data,
            op_kwargs={ ....  }

 PythonOperator.partial(
    task_id=f'....{gen_task_id}...', 
    python_callable=process_split_data, 
    op_kwargs={'processing_func': processing_func, 'aws_config': aws_config, 
    ...
    ).expand(
    templates_dict=XComArg(task)
    )

**Comment by EssKayz** [comment]
So essentially what we would _want_ to do is have the `process_split_data` callable get kwargs from the XCom return of the task, and that we have currently jerry-rigged using the templates_dict.
when we tried using expand_kwargs, the kwargs were simply passed to PythonOperator as kwargs, and we didn't find a way to pass over them to the op_kwargs.

**Comment by EssKayz** [comment]
I can try to quickly mock up a working example, if my incoherent rambling is hard to grasp - sadly not a native English speaker :disappointed:

**Comment by EssKayz** [comment]
This should be a working example, with the current working jerry-rig that we can't seem to get to work using expand_kwargs
```
dag = DAG('mockup_example',
          catchup=False,
          default_args=dag_args,
          default_view='graph',
          schedule_interval=None, dagrun_timeout=timedelta(minutes=30))


def gen_multiprocess_example(
    dag: DAG,
    gen_task_id,
    **kwargs
) -> TaskGroup:

    with TaskGroup(dag=dag, group_id=f'mockup_{gen_task_id}', prefix_group_id=False) as grp:
        task = PythonOperator(
            task_id=f'split_{gen_task_id}_data',
            python_callable=split_example,
            op_kwargs={

            }
        )

        # This works, by utilizing a dirty hack
        PythonOperator.partial(
            task_id=f'process_{gen_task_id}_data',
            python_callable=process_split_data,
            op_kwargs={

            }
        ).expand(
            templates_dict=XComArg(task)
        )

        # This doesn't work, since the kwargs get expanded to PythonOperator, not to the callable.
        # PythonOperator.partial(
        #     task_id=f'process_{gen_task_id}_data',
        #     python_callable=process_split_data,
        #     op_kwargs={

        #     }
        # ).expand_kwargs(
        #     XComArg(task)
        # ) 

    return grp


def split_example(**kwargs):
    return [{'data': '1234', 'bucket_name': 'prod'}, {'data': '4321', 'bucket_name': 'test'}]


def process_split_data(data=None, bucket_name=None, templates_dict={}, **kwargs):
    print('-' * 20)
    print('data in process_func:', data)
    print('bucket_name in process_func:', bucket_name)
    print('templates_dict in process_func:', templates_dict)
    print('-' * 20)
    print(get_data(**templates_dict, **kwargs))


def get_data(data=None, bucket_name=None, **kwargs):
    print('-' * 20)
    print('data in get_data:', data)
    print('bucket_name in get_data:', bucket_name)
    print('-' * 20)
    return data


with dag:
    tg1 = gen_multiprocess_example(
        dag,
        gen_task_id='hope_this_helps'
    )
```

**Comment by uranusjr** [comment]
Would this not work?

```python
@task(task_id=f'process_{gen_task_id}_data')
def process_split_data(data):
    ...

process_split_data.expand(data=XComArg(task))
```

**Comment by EssKayz** [comment]
Hmm, will have to check if that would work - but wouldn't it still be a good idea to try to support PythonOperators with expand_kwargs too? Not sure how complex that would be to do of course.

**Comment by uranusjr** [comment]
It is supported in concept, the main problem is actually the `op_kwargs` syntax would be pretty complex for a user to _write_ so I’m not bothering to try figuring out how it’d work.

**Comment by EssKayz** [comment]
Still not exactly sure how to get that to work, unsure if I understood the suggestion wrong, but at least done in a similar way as with PythonOperators, no task is generated to the DAG
```
from airflow.decorators import task

def get_data_from_db(table_name, **kwargs):

    @task(task_id=f'get_db_{table_name}')
    def get_db_objects(**kwargs):
        return {}


with dag:
    test_task = get_data_from_db(
        table_name='person'
    )
    test_task2 = get_data_from_db(
        table_name='employer'
    )

```

**Comment by uranusjr** [comment]
You did not call the task function. I’m going to convert this to a discussion since this is more about the gap in user knowledge instead of Airflow functionality here

**Reply by EssKayz** [reply]
In the example I posted, I tried both ways a, get_data_from_db -> which would be the task factory function. If you define an f-string in the @task annotation and have it outside of a generator function, the variable within the f-string leads to an error.
@task(task_id=f'get_db_{table_name}')
NameError: name 'table_name' is not defined

A discussion can be fine at least for now, but for our use case there doesn't seem to be a way to do things "nicely" without resorting to using pythonOperators and a dirty hack with the templates_dict for passing on kwargs through the .expand() function

To reiterate what we did accomplish using a dirty hack :
Have a task that takes in a parameter "table_name" -> generate a task export data from database into S3, 
and have another dynamically generated task that takes the data from S3, and splits it into 2...n S3 files, 
that are then passed over via .expand(tempaltes_dict=XComArg(task)) 
-> pythonOperators that get generated via .expand() each get the kwarg templates_dict, and are from there capable of processing the exported data from "table_name" table. 
Each task here gets the task_id "export_{table_name}" - "split_{table_name}" and "process_{table_name}"

**Reply by EssKayz** [reply]
obviously defining for example a loop within a dag, containing the parameters is an option, but a poor once since that would mean defining the task in each DAG separately within the loop.
